zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/user/CreateUserMultiPartitionTest.java

```diff
@@ -0,0 +1,82 @@
+
+import io.camunda.zeebe.engine.util.EngineRule;
+import io.camunda.zeebe.protocol.record.Record;
+import io.camunda.zeebe.protocol.record.RecordType;
+import io.camunda.zeebe.protocol.record.intent.CommandDistributionIntent;
+import io.camunda.zeebe.protocol.record.intent.UserIntent;
+import io.camunda.zeebe.protocol.record.value.CommandDistributionRecordValue;
+import io.camunda.zeebe.test.util.record.RecordingExporter;
+import io.camunda.zeebe.test.util.record.RecordingExporterTestWatcher;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestWatcher;
+
+public class CreateUserMultiPartitionTest {
+  private static final int PARTITION_COUNT = 3;
+
+  @ClassRule public static final EngineRule ENGINE = EngineRule.multiplePartition(PARTITION_COUNT);
+
+  @Rule public final TestWatcher testWatcher = new RecordingExporterTestWatcher();
+
+  @Test
+  public void shouldDistributeUserCreateCommand() {
+    // when
+    final var username = UUID.randomUUID().toString();
+
+    ENGINE
+        .user()
+        .newUser(username)
+        .withName("Foo Bar")
+        .withEmail("foo@bar.com")
+        .withPassword("password")
+        .create();
+
+    assertThat(
+            RecordingExporter.records()
+                .withPartitionId(1)
+                .limit(record -> record.getIntent().equals(CommandDistributionIntent.FINISHED)))
+        .extracting(
+            Record::getIntent,
+            Record::getRecordType,
+            r ->
+                // We want to verify the partition id where the deletion was distributing to and
+                // where it was completed. Since only the CommandDistribution records have a
+                // value that contains the partition id, we use the partition id the record was
+                // written on for the other records.
+                r.getValue() instanceof CommandDistributionRecordValue
+                    ? ((CommandDistributionRecordValue) r.getValue()).getPartitionId()
+                    : r.getPartitionId())
+        .startsWith(tuple(UserIntent.CREATE, RecordType.COMMAND, 1))
+        .endsWith(tuple(CommandDistributionIntent.FINISHED, RecordType.EVENT, 1));
+
+    assertThat(
+            RecordingExporter.records()
+                .withPartitionId(2)
+                .limit(record -> record.getIntent().equals(UserIntent.CREATED))
+                .collect(Collectors.toList()))
+        .extracting(Record::getIntent)
+        .containsExactly(UserIntent.CREATE, UserIntent.CREATED);
+
+    assertThat(
+            RecordingExporter.records()
+                .withPartitionId(3)
+                .limit(record -> record.getIntent().equals(UserIntent.CREATED))
+                .collect(Collectors.toList()))
+        .extracting(Record::getIntent)
+        .containsExactly(UserIntent.CREATE, UserIntent.CREATED);
```

Comment:
```suggestion
    assertThat(
            RecordingExporter.records()
                .withPartitionId(1)
                .limit(record -> record.getIntent().equals(CommandDistributionIntent.FINISHED)))
        .extracting(
            Record::getIntent,
            Record::getRecordType,
            r ->
                // We want to verify the partition id where the creation was distributing to and
                // where it was completed. Since only the CommandDistribution records have a
                // value that contains the partition id, we use the partition id the record was
                // written on for the other records.
                r.getValue() instanceof CommandDistributionRecordValue
                    ? ((CommandDistributionRecordValue) r.getValue()).getPartitionId()
                    : r.getPartitionId())
        .startsWith(
            tuple(UserIntent.CREATE, RecordType.COMMAND, 1),
            tuple(UserIntent.CREATED, RecordType.EVENT, 1),
            tuple(CommandDistributionIntent.STARTED, RecordType.EVENT, 1))
        .containsSubsequence(
            tuple(CommandDistributionIntent.DISTRIBUTING, RecordType.EVENT, 2),
            tuple(CommandDistributionIntent.ACKNOWLEDGE, RecordType.COMMAND, 2),
            tuple(CommandDistributionIntent.ACKNOWLEDGED, RecordType.EVENT, 2))
        .containsSubsequence(
            tuple(CommandDistributionIntent.DISTRIBUTING, RecordType.EVENT, 3),
            tuple(CommandDistributionIntent.ACKNOWLEDGE, RecordType.COMMAND, 3),
            tuple(CommandDistributionIntent.ACKNOWLEDGED, RecordType.EVENT, 3))
        .endsWith(tuple(CommandDistributionIntent.FINISHED, RecordType.EVENT, 1));

    for (int partitionId = 2; partitionId < PARTITION_COUNT; partitionId++) {
      assertThat(
              RecordingExporter.records()
                  .withPartitionId(partitionId)
                  .limit(record -> record.getIntent().equals(UserIntent.CREATED))
                  .collect(Collectors.toList()))
          .extracting(Record::getIntent)
          .containsExactly(UserIntent.CREATE, UserIntent.CREATED);
    }
```

❌ I've expanded on the assertions here. We should make sure each step of the distribution works as expected, and in the correct order. We need to make sure the other partitions actually acknowledge the distribution, and whilst theoretically just checking for the `FINISHED` should do, this is more easy to understand when this test case fails, as the message points directly to a missed acknowledgement.

I also put the user create and created assertion in a for-loop. It's not that significant but reduces some code duplication.

Reply:
Ah thank you for this context, I agree it definitely makes sense to check all the steps